zookeeper/core/component.py

```diff
@@ -0,0 +1,550 @@
+        if isinstance(field, EmptyField):
+            instance.__component_fields__[name] = InheritedField(
+                annotated_type=field.annotated_type, is_overriden=False
+            )
+    else:
+        object.__setattr__(instance, name, value)
+        instance.__component_fields__[name] = NonEmptyField(
+            annotated_type=field.annotated_type, is_overriden=False
+        )
+
+
+####################################
+# Component class method wrappers. #
+####################################
+
+
+def init_wrapper(init_fn):
+    # Components need to be instantiable without arguments, so check that
+    # `init_fn` does not accept any positional arguments without default values.
+    for i, (name, param) in enumerate(inspect.signature(init_fn).parameters.items()):
+        if (
+            i > 0
+            and param.default == inspect.Parameter.empty
+            and param.kind
+            not in [inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD]
+        ):
+            raise ValueError(
+                "The `__init__` method of a component must not accept any "
+                "positional arguments, as the component configuration process "
+                "requires component classes to be instantiable without arguments."
+            )
+
+    def __component_init__(instance, **kwargs):
+        # Fake 'super' call.
+        if init_fn != object.__init__:
+            init_fn(instance, **kwargs)
+
+        instance.__component_fields__ = {}
+
+        # Populate `__component_fields__` with all annotations set on this class
+        # and all superclasses. We have to go through the MRO chain and collect
+        # them in reverse order so that they are correctly overriden.
+        annotations = {}
+        for base_class in reversed(inspect.getmro(instance.__class__)):
+            annotations.update(getattr(base_class, "__annotations__", {}))
+        instance.__component_fields__ = {}
+        for name, annotated_type in annotations.items():
+            if name in object.__dir__(instance):  # type: ignore
+                instance.__component_fields__[name] = NonEmptyField(
+                    annotated_type, is_overriden=False
+                )
+            else:
+                instance.__component_fields__[name] = EmptyField(annotated_type)
+
+        if init_fn == object.__init__:
+            for name, value in kwargs.items():
+                if name in instance.__component_fields__:
+                    set_field_value(instance, name, value)
+                else:
+                    raise ValueError(
+                        f"Argument '{name}' does not correspond to any annotated field "
+                        f"of '{type_name_str(instance.__class__)}'."
+                    )
+
+    return __component_init__
+
+
+def dir_wrapper(dir_fn):
+    def __component_dir__(instance):
+        return set(dir_fn(instance)).union(set(instance.__component_fields__.keys()))
```

Comment:
I think that would be a bit more readable:
```suggestion
        return set(dir_fn(instance)) | set(instance.__component_fields__.keys())
```